File: padroes-de-projeto/relogio-singleton/date.py
class Date:
    _day = _month = _year = None

    def __init__(self, day: int, month: int, year: int):
        self.day = day
        self.year = year
        self.month = month

    @property
    def day(self) -> int:
        return self._day

    @day.setter
    def day(self, value: int):
        if value > 31 or value < 1:
            raise ValueError('invalid date.')
        self._day = int(value)

    @property
    def month(self) -> int:
        return self._month

    @month.setter
    def month(self, value: int):
        if value > 12 or value < 1:
            raise ValueError('invalid date.')
        if value in (4, 6, 9, 11) and self.day > 30:
            raise ValueError('invalid date.')
        if value == 2 and self.day > 29:
            raise ValueError('invalid date.')
        if not self.is_leap_year(self.year) and self.day > 28:
            raise ValueError('invalid date.')
        self._month = int(value)

    @property
    def year(self) -> int:
        return self._year

    @year.setter
    def year(self, value: int):
        if value < 1:
            raise ValueError('invalid date.')
        self._year = int(value)

    @staticmethod
    def is_leap_year(year: int) -> bool:
        return (year % 4 == 0 and year % 100 != 0) or year % 400 == 0

    def __repr__(self):
        return f'{self.day:0>2}/{self.month:0>2}/{self.year:0>4}'

File: programacao-web/crud-user/src/crud/Date.java
package crud;

public class Date {

    private int day;
    private int month;
    private int year;

    private int[] months30days = {4, 6, 9, 11};

    public Date(int day, int month, int year) {
        this.setDay(day);
        this.setYear(year);
        this.setMonth(month);
    }

    public int getDay() {
        return this.day;
    }

    public void setDay(int day) throws IllegalArgumentException {
        if (day < 1 | day > 31) {
            throw new IllegalArgumentException("invalid date");
        }
        this.day = day;
    }

    public int getMonth() {
        return this.month;
    }

    public void setMonth(int month) throws IllegalArgumentException {
        if (month < 1 || month > 12) {
            throw new IllegalArgumentException("invalid date");
        }
        if (Operand.in(month, this.months30days) && this.getDay() > 30) {
            throw new IllegalArgumentException("invalid date");
        }
        if (month == 2 && this.getDay() > 29) {
            throw new IllegalArgumentException("invalid date");
        }
        if (!isLeapYear(this.getYear()) && this.getDay() > 28) {
            throw new IllegalArgumentException("invalid date");
        }
        this.month = month;
    }

    public int getYear() {
        return this.year;
    }

    public void setYear(int year) throws IllegalArgumentException{
        if (year < 1) {
            throw new IllegalArgumentException("invalid date");
        }
        this.year = year;
    }

    public static boolean isLeapYear(int year) {
        return (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
    }

    public String toString() {
        return this.getDay() + "/" + this.getMonth() + "/" + this.getYear();
    }
}

class Operand {
	public static boolean in(int value, int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == value) {
                return true;
            }
        }
        return false;
    }
}

